RCNN_tf.py

```diff
@@ -0,0 +1,338 @@
+        name='drop_2'
+    )
+
+    # Recurrent convolutional layer
+    rcl_layer_3_ = rcl(
+        input_data=pooling_2_,
+        num_input_channels=num_filter_,
+        filter_shape=(3, 3),
+        num_filter=num_filter_,
+        name='rcl_layer_3'
+    )
+
+    dropout_3_ = tf.nn.dropout(
+        rcl_layer_3_,
+        rate=rate_placeholder_,
+        name='drop_3'
+    )
+
+    rcl_layer_4_ = rcl(
+        input_data=dropout_3_,
+        num_input_channels=num_filter_,
+        filter_shape=(3, 3),
+        num_filter=num_filter_,
+        name='rcl_layer_4'
+    )
+
+    # Max pooling layer
+    global_max_ = global_max_pooling_layer(rcl_layer_4_)
+
+    # Flatten tensor for softmax layer
+    flatten_ = tf.reshape(global_max_, [-1, global_max_.shape[-1].value])
+
+    # Softmax layer
+    result_, linear_trans_ = softmax_layer(
+        flatten_,
+        global_max_.shape[-1].value,
+        output_shape_[1]
+    )
+
+    # cross entropy and accuracy matrix
+    cross_entropy_ = tf.reduce_mean(
+        tf.nn.softmax_cross_entropy_with_logits_v2(logits=linear_trans_, labels=output_placeholder_)
+    )
+
+    optimiser_ = tf.train.AdamOptimizer(learning_rate=learning_rate_).minimize(cross_entropy_)
+
+    # Accuracy
+    accuracy_ = accuracy(
+        output_placeholder_,
+        result_
+    )
+
+    init_op_ = tf.global_variables_initializer()
+
+    with tf.Session() as sess_:
+        # initialise the variables
+        sess_.run(init_op_)
+        total_batch_ = int(training_data_.shape[0] / batch_size_)
+        for epoch in range(epochs_):
+            avg_cost_ = 0
+            for i in range(total_batch_):
+                progress = (i / float(total_batch_-1)) * 100
+                print '\r {:.1f}%'.format(progress), '\t{0}> '.format('#' * int(progress)),
+                batch_x_ = training_data_[i * batch_size_: (i + 1) * batch_size_]
+                batch_y_ = training_labels_[i * batch_size_: (i + 1) * batch_size_]
+                _, cost_ = sess_.run([optimiser_, cross_entropy_],
+                                     feed_dict={
+                                         input_placeholder_: batch_x_,
+                                         output_placeholder_: batch_y_,
+                                         rate_placeholder_: 0.2
+                                     })
+                avg_cost_ += cost_ / total_batch_
+
+            test_acc_ = sess_.run(accuracy_,
+                                  feed_dict={
+                                      input_placeholder_: test_data_,
+                                      output_placeholder_: test_labels_,
+                                      rate_placeholder_: 0
+                                  })
+            print "\nEpoch:", (epoch + 1), \
+                "cost =", "{:.3f}".format(avg_cost_),\
+                "test accuracy: {:.3f}".format(test_acc_)
+
+        print "\nTraining complete!"
+        print sess_.run(accuracy_,
```

Comment:
this can really be simplified using `tf.data.Dataset` which can handle shuffling, batching, repeating etc.

Reply:
I will take a look on that